Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx

```diff
@@ -0,0 +1,304 @@
+static std::array<int, 64> refManu2ds_st345_v2 = {
+  62, 61, 63, 60, 59, 55, 58, 57, 56, 54, 50, 46, 42, 39, 37, 41,
+  35, 36, 33, 34, 32, 38, 43, 40, 45, 44, 47, 48, 49, 52, 51, 53,
+  7, 6, 5, 4, 2, 3, 1, 0, 9, 11, 13, 15, 17, 19, 21, 23,
+  31, 30, 29, 28, 27, 26, 25, 24, 22, 20, 18, 16, 14, 12, 10, 8};
+
+#define refManu2ds_st345 refManu2ds_st345_v5
+
+// inverse channel conversion matrix
+static std::array<int, 64> refDs2manu_st345;
+
+// function returning the RUN3 DualSAMPA channel number given the original RUN2 channel
+static int manu2ds(int i)
+{
+  return refManu2ds_st345[i];
+}
+
+// function returning the original RUN2 channel number given the RUN3 DualSAMPA channel
+static int ds2manu(int i)
+{
+  return refDs2manu_st345[i];
+}
+
+static void patchPage(gsl::span<const std::byte> rdhBuffer, bool verbose)
+{
+  static int mNrdhs = 0;
+  auto& rdhAny = *reinterpret_cast<RDH*>(const_cast<std::byte*>(&(rdhBuffer[0])));
+  mNrdhs++;
+
+  auto cruId = o2::raw::RDHUtils::getCRUID(rdhAny) & 0xFF;
+  auto flags = o2::raw::RDHUtils::getCRUID(rdhAny) & 0xFF00;
+  auto endpoint = o2::raw::RDHUtils::getEndPointID(rdhAny);
+  uint32_t feeId = cruId * 2 + endpoint + flags;
+  o2::raw::RDHUtils::setFEEID(rdhAny, feeId);
+
+  if (verbose) {
+    std::cout << mNrdhs << "--\n";
+    o2::raw::RDHUtils::printRDH(rdhAny);
+  }
+};
+
+//=======================
+// Data decoder
+
+DataDecoder::DataDecoder(SampaChannelHandler channelHandler, RdhHandler rdhHandler, std::string mapCRUfile, std::string mapFECfile, bool ds2manu, bool verbose)
+  : mChannelHandler(channelHandler), mRdhHandler(rdhHandler), mMapCRUfile(mapCRUfile), mMapFECfile(mapFECfile), mDs2manu(ds2manu), mPrint(verbose)
+{
+  init();
+}
+
+static bool isValidDeID(int deId)
+{
+  for (auto id : deIdsForAllMCH) {
+    if (id == deId) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
+{
+  size_t ndigits{0};
+
+  uint32_t orbit;
+
+  auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+    if (mChannelHandler) {
+      mChannelHandler(dsElecId, channel, sc);
+    }
+
+    if (mDs2manu) {
+      channel = ds2manu(int(channel));
+    }
+
+    uint32_t digitadc = sc.sum();
+
+    int deId{-1};
+    int dsIddet{-1};
+    if (auto opt = mElec2Det(dsElecId); opt.has_value()) {
+      DsDetId dsDetId = opt.value();
+      dsIddet = dsDetId.dsId();
+      deId = dsDetId.deId();
+    }
+    if (mPrint) {
+      auto s = asString(dsElecId);
+      auto ch = fmt::format("{}-CH{:02d}", s, channel);
+      std::cout << ch << "  "
+                << "deId " << deId << "  dsIddet " << dsIddet << std::endl;
+    }
+
+    if (deId < 0 || dsIddet < 0 || !isValidDeID(deId)) {
+      return;
+    }
+
+    int padId = -1;
+    const Segmentation& segment = segmentation(deId);
+
+    padId = segment.findPadByFEE(dsIddet, int(channel));
+    if (mPrint) {
```

Comment:
Generally I would agree, however at least in this specific case the log message requires building a string from the electronics mapping data, and I would prefer to avoid this extra string creation for normal, non-verbose execution.

In fact, this output is really only needed for debugging, and I am changing the code such that it can be enabled/disabled via a command-line option...

Reply:
ok